### PR TITLE
Fixing code of tau embedding (TauAnalysis/MCEmbeddingTools) for run2 needs

### PR DESCRIPTION
**Pull request by swayand**
This is a follow up PR of https://github.com/cms-sw/cmssw/pull/16271

This PR replaces https://github.com/cms-sw/cmssw/pull/16633

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
I'm posting a copy of my signoff message from the 81X precursor PR #16271 

for https://github.com/cms-sw/cmssw/pull/17095 301c873

- reco review was added to aid with this analysis PR review due to a strong coupling with the reco software algorithm setup. 
    - Code changes look OK within the analysis scope of the MCEmbeddingTools.
- jenkins tests pass including the unit test running this workflow. There are no changes in the matrix tests.

See also some notes in https://github.com/cms-sw/cmssw/pull/16271#issuecomment-267666417 . While the detailed embedding is useful as a reference, I think that there should be a faster standard methodology relying on at least AOD to do embedding, especially in view of increasing pileup [the slow SIMembedding step could even be done with fastSim tools].

**Comment by civanch** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by slava77** [comment]
ping
analysis and generators signatures are missing. 
It would be nice to converge before the PR is more than 3 months old (81X "parent" was submitted Oct 19).

**Comment by monttj** [comment]
+1

**Comment by perrozzi** [comment]
+1

File: GeneratorInterface/Core/interface/BaseHadronizer.h [hunk 1]
```diff
@@ -111,6 +111,8 @@ namespace gen {
     std::vector<double> randomInitWeights_;
     std::vector<std::string> randomInitConfigDescriptions_;
     std::vector<std::string> gridpackPaths_;
+
+    char *args[5];
```
Comment (davidlange6) [inline GeneratorInterface/Core/interface/BaseHadronizer.h]: Hi @swayand- why is this change needed? It looks like this variable is used in only one spot.

File: TauAnalysis/MCEmbeddingTools/plugins/DYToMuMuGenFilter.cc [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+			//std::cout << "pdgId" << gen_particle.pdgId() << std::endl;
+			//std::cout << "nDau" << gen_particle.numberOfDaughters() << std::endl;
+			//std::cout << "Dau1" << gen_particle->daughters->at(0).pdgId() << std::endl;
+			//std::cout << "Dau2" << gen_particle.numberOfDaughters() << std::endl;
+			//std::cout << "Dau1 " << gen_particle.daughter(0)->pdgId() << std::endl;
+			//std::cout << "Dau2 " << gen_particle.daughter(1)->pdgId() << std::endl;
+			//std::cout << gen_particle.daughter(1)->pdgId()+gen_particle.daughter(0)->pdgId() << std::endl;
+			
+			// Check if daugther particles are muons
+		  if (abs(gen_particle.daughter(0)->pdgId()) == 13
```
Comment (davidlange6) [inline TauAnalysis/MCEmbeddingTools/plugins/DYToMuMuGenFilter.cc]: please use std::abs here.

**Comment by swayand** [comment]
On 01/18/2017 11:25 AM, David Lange wrote:
>
> *@davidlange6* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In GeneratorInterface/Core/interface/BaseHadronizer.h 
> <https://github.com/cms-sw/cmssw/pull/17095#pullrequestreview-17189412>:
>
> > @@ -111,6 +111,8 @@ namespace gen {
>       std::vector<double> randomInitWeights_;
>       std::vector<std::string> randomInitConfigDescriptions_;
>       std::vector<std::string> gridpackPaths_;
> +
> +    char *args[5];
>
> Hi @swayand- why is this change needed? It looks like this variable is 
> used in only one spot.
>
> —
>
*@davidlange6*

I made this change because of the static code analyzer gives the 
previous implementation several critical bugs. Therefore i fixed it, 
even it has nothing to do with the embedding code itself.

**Comment by swayand** [comment]
@davidlange6 

i made the requested fixes (added std to the abs function)

**Comment by civanch** [comment]
please test

File: TauAnalysis/MCEmbeddingTools/plugins/DYToMuMuGenFilter.cc [hunk 3]
```diff
@@ -74,9 +74,9 @@ bool DYToMuMuGenFilter::filter(edm::Event& iEvent, const edm::EventSetup& iSetup
 			//std::cout << gen_particle.daughter(1)->pdgId()+gen_particle.daughter(0)->pdgId() << std::endl;
 			
 			// Check if daugther particles are muons
-		  if (abs(gen_particle.daughter(0)->pdgId()) == 13  
-		      && fabs(gen_particle.daughter(0)->eta())<2.6  
-		      && fabs(gen_particle.daughter(1)->eta())<2.6
+		  if (std::abs(gen_particle.daughter(0)->pdgId()) == 13  
+		      && std::fabs(gen_particle.daughter(0)->eta())<2.6  
+		      && std::fabs(gen_particle.daughter(1)->eta())<2.6
```
Comment (slava77) [inline TauAnalysis/MCEmbeddingTools/plugins/DYToMuMuGenFilter.cc]: better use std::abs here as well

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by perrozzi** [comment]
+1

**Comment by swayand** [comment]
@monttj 
here still the analysis flag is missing. Could you check again, the changes to last time was that the abs/fabs was changed to std::abs.

**Comment by davidlange6** [comment]
Hi @swayand - i'm not really waiting on @monttj - rather there are some unanswered questions to you (maybe just one) and I just notice that this PR no longer merges, so needs to be rebased.

**Comment by swayand** [comment]
@davidlange6  could you pleas point me to the unanswered question. 

Afterward i will do the rebase, or resolves the conflicts

**Comment by slava77** [comment]
@davidlange6 
https://github.com/cms-sw/cmssw/pull/17095#issuecomment-277953790

```
@davidlange6 could you pleas point me to the unanswered question.

Afterward i will do the rebase, or resolves the conflicts
```

**Comment by davidlange6** [comment]
ok, you answered it but in a different part of the PR

File: GeneratorInterface/Core/src/BaseHadronizer.cc [hunk 4]
```diff
@@ -104,7 +104,6 @@ const std::vector<std::string> BaseHadronizer::theSharedResources;
     edm::FileInPath script("GeneratorInterface/LHEInterface/data/run_generic_tarball_cvmfs.sh");
     const char *outfilename = "cmsgrid_final.lhe";
     
-    char *args[5];
```
Comment (davidlange6) [inline GeneratorInterface/Core/src/BaseHadronizer.cc]: I think a c++ way that avoids newing char*s for this is just

std::array<std::string,4>argStrs;
argStrs[0]=script.fullPath();
argStrs[1]=gridpackPath();
argStrs[2]=nevStream.str();
argStrs[3]=randomStream.str();
std::array<char*,5>args = { &argStrs[0][0], &argStrs[1][0],&argStrs[2][0],&argStrs[3][0], NULL};

then memory is managed for you

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by swayand** [comment]
Hi, 

i have updated the PR. There are three changes between the previous version. 

1. Have updated the EmbeddingLHEProducer, because there is a bug in pyhtia8, which don't restore the tau life time. Therefore i have to add also the HepRandomEngine. The solution was already tested and works. I also contact the pythia8 team and reported about this issue.
https://github.com/swayand/cmssw/blob/4fa62e12dedb8d0f38db233ae76bce2ba810cd42/TauAnalysis/MCEmbeddingTools/plugins/EmbeddingLHEProducer.cc#L297-L300

2. I also made small code improvements, where i replaced one new allocator (where the memory wasn't freed), by just calling the function 
https://github.com/swayand/cmssw/blob/c6ee7f5d5761cabdb44025d3d1bb70600805a954/TauAnalysis/MCEmbeddingTools/plugins/EmbeddingVertexCorrector.cc#L126-L127

3. Then I implemented the suggestions from @davidlange6 and implemented std::arrays<char*> instead of char*[] 
https://github.com/swayand/cmssw/blob/f98e6e038ff93c5217f683424ebdcc616bc4aae0/GeneratorInterface/Core/src/BaseHadronizer.cc#L109-L115

**Comment by civanch** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Comment by Dr15Jones** [comment]
@davidlange6 this fails to link in the IB
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc630/CMSSW_9_0_X_2017-02-17-2300/TauAnalysis/MCEmbeddingTools

**Comment by davidlange6** [comment]
right- thats what I attempted to fix for the 02-18-2300 Ib...
